cpp/src/spatial/linestring_distance.cu

```diff
@@ -0,0 +1,335 @@
+
+namespace cuspatial {
+namespace {
+
+/** @brief Get the index that is one-past the end point of linestring at @p linestring_idx
+ *
+ * @note The last endpoint of the linestring is not included in the offset array, thus
+ * @p num_points is returned.
+ */
+template <typename SizeType, typename OffsetIterator>
+inline SizeType __device__
+endpoint_index_of_linestring(SizeType const& linestring_idx,
+                             OffsetIterator const& linestring_offsets_begin,
+                             SizeType const& num_linestrings,
+                             SizeType const& num_points)
+{
+  return (linestring_idx == (num_linestrings - 1)
+            ? (num_points)
+            : *(linestring_offsets_begin + linestring_idx + 1)) -
+         1;
+}
+
+/**
+ * @brief Computes shortest distance between @p C and segment @p A @p B
+ */
+template <typename T>
+T __device__ point_to_segment_distance_squared(vec_2d<T> const& c,
+                                               vec_2d<T> const& a,
+                                               vec_2d<T> const& b)
+{
+  vec_2d<T> ab   = b - a;
+  auto ac        = c - a;
+  auto bc        = c - b;
+  auto l_squared = dot(ab, ab);
+  if (l_squared == 0) { return dot(ac, ac); }
+  auto r = dot(ac, ab) / l_squared;
+  if (r <= 0 or r >= 1) { return std::min(dot(ac, ac), dot(bc, bc)); }
+  auto p  = a + r * ab;
+  auto pc = c - p;
+  return dot(pc, pc);
+}
+
+/**
+ * @brief Computes shortest distance between two segments that doesn't intersect.
+ */
+template <typename T>
+T __device__ segment_distance_no_intersect_or_colinear(vec_2d<T> const& a,
+                                                       vec_2d<T> const& b,
+                                                       vec_2d<T> const& c,
+                                                       vec_2d<T> const& d)
+{
+  auto dist_sqr = std::min(std::min(point_to_segment_distance_squared(a, c, d),
+                                    point_to_segment_distance_squared(b, c, d)),
+                           std::min(point_to_segment_distance_squared(c, a, b),
+                                    point_to_segment_distance_squared(d, a, b)));
+  return dist_sqr;
+}
+
+/**
+ * @brief Computes shortest distance between two segments.
+ *
+ * If two segments intersect, the distance is 0. Otherwise compute the shortest point
+ * to segment distance.
+ */
+template <typename T>
+T __device__ squared_segment_distance(vec_2d<T> const& a,
+                                      vec_2d<T> const& b,
+                                      vec_2d<T> const& c,
+                                      vec_2d<T> const& d)
+{
+  auto ab    = b - a;
+  auto ac    = c - a;
+  auto cd    = d - c;
+  auto denom = det(ab, cd);
+
+  if (denom == 0) {
+    // Segments parallel or collinear
+    return segment_distance_no_intersect_or_colinear(a, b, c, d);
+  }
+  auto r_numer = det(ac, cd);
+  auto r       = r_numer / denom;
+  auto s       = det(ac, ab) / denom;
```

Comment:
Could always precompute 1/denom and then multiply, but I'm probably microoptimizing here.

Reply:
Out of academic curiosity, are there precision related concerns when replacing division with multiply by reciprocal?

Reply:
That's a great question. Yes, there will be differences. Division is usually implemented using a root finding scheme. The numerical convergence properties of that scheme will result in different results if you do `z = y/x` vs doing `tmp = 1 / x; z = y * x`. The bigger issue will be truncation error resulting from the magnitude of `x`. The difference in numerical error between division and multiplication by reciprocal will probably be dominated by the same kind of thing as doing summations of a sequence: you will lose more precision when you perform operations between numbers of very different scales. In this example, if x is of order unity then `1/x` will be computed precisely and you won't lose much, whereas if `x` is much larger or much smaller than `1` then you'll lose some digits. Which approach is more precise will depend on whether `x` and `y` are of similar magnitude. If `x~y` but not `x~1` (which is probably a realistic expectation) you'll get more precise results from doing `z = y/x` than storing the reciprocal as the intermediate. 

All of this is to say that I don't know enough about what tradeoffs cuspatial wants to/is willing to make. My guess would be that choosing the more precise result may be the better solution unless and until someone really wants to squeeze every last drop of performance out of this code. I don't think we're at the point where we try to optimize that far in cuspatial.

Reply:
Here is an amazingly detailed discussion of this for x86. https://stackoverflow.com/a/45899202/749748  This of course will be somewhat different for NVIDIA / SASS. 

Like x86, our GPUs do have an rcp instruction which can be used for approximate division. There are multiple other ways to divide other than `/`, including `__fdividef(x,y)`. There is a bit of discussion of the accuracy of division in the CUDA Programming guide: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#intrinsic-functions

All this said, in the absence of an accuracy test specific to real data and a corresponding benchmark, I think you should stick to the division you have in the code above, since it is guaranteed to be IEEE compliant and has the best chance of matching results of other libraries. 

For the future, the fact that these computed values are only used for comparison to zero and one may provide optimization opportunities. But right now let's just make it robust.

Reply:
Agreed, this kind of microoptimization really shouldn't happen without a compelling benchmark as well as validation that the potential loss of precision wouldn't be a problem.

Reply:
@isvoid looks like you already changed to computing reciprocal. Have you compared accuracy or performance vs. previous? It may not make much difference to accuracy as I said, since this is only used for comparing if the results are between 0 and 1.

Reply:
Short answer is I haven't done any specific test to this yet. For accuracy the current implementation passes all existing tests (including offline geolife tests) the same as previous code.

Reply:
On to performance, with the offline dataset (8800 pairs of linestrings), there is a visible 6% speedup with the multiply by reciprocal approach. The below timing are acquired by timing the python bindings (with different cpp implementation) with `timeit`. Each ran 50 times, the average timings are:
```
0.7526701703201979 for multiply by reciprocal
0.8014753964729607 for division
```

Reply:
That's good!

Reply:
Well then I guess this was a worthwhile optimization to suggest after all!